src/library/scala/collection/immutable/TreeMap.scala

```diff
@@ -27,6 +29,37 @@ object TreeMap extends ImmutableSortedMapFactory[TreeMap] {
   def empty[A, B](implicit ord: Ordering[A]) = new TreeMap[A, B]()(ord)
   /** $sortedMapCanBuildFromInfo */
   implicit def canBuildFrom[A, B](implicit ord: Ordering[A]): CanBuildFrom[Coll, (A, B), TreeMap[A, B]] = new SortedMapCanBuildFrom[A, B]
+  private val legacySerialisation = System.getProperty("scala.collection.immutable.TreeMap.newSerialisation", "false") != "false"
+
+  private class TreeMapProxy[A, B](
+    @transient private[this] var tree: RedBlackTree.Tree[A, B],
+    @transient private[this] var ordering: Ordering[A]) extends Serializable {
+
+    @throws[IOException]
+    private[this] def writeObject(out: java.io.ObjectOutputStream) = {
+      out.writeInt(RB.count(tree))
+      out.writeObject(ordering)
+      RB.foreachEntry(tree, {
+        (k: A, v: B) =>
+          out.writeObject(k)
+          out.writeObject(v)
+      })
+    }
+    @throws[IOException]
+    private[this] def readObject(in: java.io.ObjectInputStream) = {
+      val size = in.readInt()
+      ordering = in.readObject().asInstanceOf[Ordering[A]]
+      implicit val ord = ordering
+      for (i <- 0 until size) {
+        val key = in.readObject().asInstanceOf[A]
+        val value = in.readObject().asInstanceOf[B]
+        RB.update(tree, key, value, overwrite = true)
```

Comment:
Since you know the size upfront, there's an opportunity to build the tree much faster. We already assume that the ordering is stable so you can rely on the elements to be in order. You could add a sanity check and fall back to unordered building if it fails but I wouldn't bother.

Reply:
Hi @szeiger you assume more knowledge than I have on how to build the ordered tre. If you can explain then the other PRs that i have can benefit from that as well

Reply:
I implemented a suitable algorithm for 2.13 here: https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/immutable/RedBlackTree.scala#L418. The encoding of the red-black trees hasn't changed, this should work equally well in 2.12.

Reply:
tried backporting RedBlackTree to 2.12 to see what breaks - https://github.com/scala/scala/pull/8749

Reply:
rebuild the PR on top of https://github.com/scala/scala/pull/8749

added improved tree building in https://github.com/scala/scala/pull/8747/commits/afe09b3d5ddeadab3aeda04c92e221dc6729aa3a